File: backend/src/routes/email.ts
import { Router, Request, Response } from 'express';
import { emailService } from '../services/emailService';

const router = Router();

// Get all email configurations
router.get('/configs', (req: Request, res: Response) => {
  try {
    const configs = emailService.getAllConfigs();
    res.json({ success: true, data: configs });
  } catch (error) {
    res.status(500).json({ success: false, message: '获取邮箱配置失败', error: String(error) });
  }
});

// Create new email configuration
router.post('/configs', async (req: Request, res: Response) => {
  try {
    const { email, imap_host, imap_port, password, is_active } = req.body;
    
    if (!email || !imap_host || !password) {
      return res.status(400).json({ success: false, message: '邮箱地址、IMAP服务器和密码是必填项' });
    }

    // Test connection first
    const testResult = await emailService.testConnection({
      email,
      imap_host,
      imap_port: imap_port || 993,
      password
    });

    if (!testResult.success) {
      return res.status(400).json({ success: false, message: testResult.message });
    }

    const config = emailService.createConfig({
      email,
      imap_host,
      imap_port: imap_port || 993,
      password,
      is_active: is_active !== undefined ? is_active : 1
    });

    res.status(201).json({ 
      success: true, 
      data: { ...config, password: '********' }, 
      message: '邮箱配置创建成功' 
    });
  } catch (error) {
    res.status(500).json({ success: false, message: '创建邮箱配置失败', error: String(error) });
  }
});

// Update email configuration
router.put('/configs/:id', (req: Request, res: Response) => {
  try {
    const updated = emailService.updateConfig(req.params.id, req.body);
    if (!updated) {
      return res.status(404).json({ success: false, message: '邮箱配置不存在或更新失败' });
    }
    res.json({ success: true, message: '邮箱配置更新成功' });
  } catch (error) {
    res.status(500).json({ success: false, message: '更新邮箱配置失败', error: String(error) });
  }
});

// Delete email configuration
router.delete('/configs/:id', (req: Request, res: Response) => {
  try {
    const deleted = emailService.deleteConfig(req.params.id);
    if (!deleted) {
      return res.status(404).json({ success: false, message: '邮箱配置不存在' });
    }
    res.json({ success: true, message: '邮箱配置删除成功' });
  } catch (error) {
    res.status(500).json({ success: false, message: '删除邮箱配置失败', error: String(error) });
  }
});

// Test email connection
router.post('/test', async (req: Request, res: Response) => {
  try {
    const { email, imap_host, imap_port, password } = req.body;
    
    if (!email || !imap_host || !password) {
      return res.status(400).json({ success: false, message: '请提供完整的连接信息' });
    }

    const result = await emailService.testConnection({
      email,
      imap_host,
      imap_port: imap_port || 993,
      password
    });

    res.json({ success: result.success, message: result.message });
  } catch (error) {
    res.status(500).json({ success: false, message: '测试连接失败', error: String(error) });
  }
});

// Get email logs
router.get('/logs', (req: Request, res: Response) => {
  try {
    const { configId, limit } = req.query;
    const logs = emailService.getLogs(
      configId as string, 
      limit ? parseInt(limit as string) : 50
    );
    res.json({ success: true, data: logs });
  } catch (error) {
    res.status(500).json({ success: false, message: '获取邮件日志失败', error: String(error) });
  }
});

// Start monitoring for a specific config
router.post('/monitor/start/:id', (req: Request, res: Response) => {
  try {
    const started = emailService.startMonitoring(req.params.id);
    if (!started) {
      return res.status(400).json({ success: false, message: '无法启动监控，请检查配置' });
    }
    res.json({ success: true, message: '邮箱监控已启动' });
  } catch (error) {
    res.status(500).json({ success: false, message: '启动监控失败', error: String(error) });
  }
});

// Stop monitoring for a specific config
router.post('/monitor/stop/:id', (req: Request, res: Response) => {
  try {
    const stopped = emailService.stopMonitoring(req.params.id);
    res.json({ success: true, message: stopped ? '邮箱监控已停止' : '监控未在运行' });
  } catch (error) {
    res.status(500).json({ success: false, message: '停止监控失败', error: String(error) });
  }
});

// Get monitoring status
router.get('/monitor/status', (req: Request, res: Response) => {
  try {
    const status = emailService.getMonitoringStatus();
    res.json({ success: true, data: status });
  } catch (error) {
    res.status(500).json({ success: false, message: '获取监控状态失败', error: String(error) });
  }
});

// Manual check for new emails
router.post('/check/:id', async (req: Request, res: Response) => {
  try {
    const result = await emailService.manualCheck(req.params.id);
    res.json({ 
      success: result.success, 
      message: result.message,
      data: { newEmails: result.newEmails }
    });
  } catch (error) {
    res.status(500).json({ success: false, message: '检查邮件失败', error: String(error) });
  }
});

export default router;

File: backend-go/internal/handlers/email.go
package handlers

import (
	"errors"
	"fmt"
	"strconv"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
	"smart-bill-manager/internal/middleware"
	"smart-bill-manager/internal/services"
	"smart-bill-manager/internal/utils"
)

type EmailHandler struct {
	emailService *services.EmailService
}

func NewEmailHandler(emailService *services.EmailService) *EmailHandler {
	return &EmailHandler{emailService: emailService}
}

func (h *EmailHandler) RegisterRoutes(r *gin.RouterGroup) {
	r.GET("/configs", h.GetAllConfigs)
	r.POST("/configs", h.CreateConfig)
	r.PUT("/configs/:id", h.UpdateConfig)
	r.DELETE("/configs/:id", h.DeleteConfig)
	r.POST("/test", h.TestConnection)
	r.GET("/logs", h.GetLogs)
	r.DELETE("/logs", h.ClearLogs)
	r.POST("/logs/:id/parse", h.ParseLog)
	r.GET("/logs/:id/export", h.ExportLogEML)
	r.POST("/monitor/start/:id", h.StartMonitoring)
	r.POST("/monitor/stop/:id", h.StopMonitoring)
	r.GET("/monitor/status", h.GetMonitoringStatus)
	r.GET("/status", h.GetMonitoringStatus) // Alias
	r.POST("/check/:id", h.ManualCheck)
}

func (h *EmailHandler) GetAllConfigs(c *gin.Context) {
	ctx, cancel := withReadTimeout(c)
	defer cancel()

	configs, err := h.emailService.GetAllConfigsCtx(ctx, middleware.GetEffectiveUserID(c))
	if err != nil {
		if handleReadTimeoutError(c, err) {
			return
		}
		utils.Error(c, 500, "获取邮箱配置失败", err)
		return
	}

	utils.SuccessData(c, configs)
}

func (h *EmailHandler) CreateConfig(c *gin.Context) {
	var input services.CreateEmailConfigInput
	if err := c.ShouldBindJSON(&input); err != nil {
		utils.Error(c, 400, "邮箱地址、IMAP服务器和密码是必填项", err)
		return
	}

	// Test connection first
	success, message := h.emailService.TestConnection(input.Email, input.IMAPHost, input.IMAPPort, input.Password)
	if !success {
		utils.Error(c, 400, message, nil)
		return
	}

	config, err := h.emailService.CreateConfig(middleware.GetEffectiveUserID(c), input)
	if err != nil {
		utils.Error(c, 500, "创建邮箱配置失败", err)
		return
	}

	utils.Success(c, 201, "邮箱配置创建成功", config.ToResponse())
}

func (h *EmailHandler) UpdateConfig(c *gin.Context) {
	id := c.Param("id")
	var data map[string]interface{}
	if err := c.ShouldBindJSON(&data); err != nil {
		utils.Error(c, 400, "参数错误", err)
		return
	}

	if err := h.emailService.UpdateConfig(middleware.GetEffectiveUserID(c), id, data); err != nil {
		utils.Error(c, 404, "邮箱配置不存在或更新失败", err)
		return
	}

	utils.Success(c, 200, "邮箱配置更新成功", nil)
}

func (h *EmailHandler) DeleteConfig(c *gin.Context) {
	id := c.Param("id")
	if err := h.emailService.DeleteConfig(middleware.GetEffectiveUserID(c), id); err != nil {
		utils.Error(c, 404, "邮箱配置不存在", nil)
		return
	}

	utils.Success(c, 200, "邮箱配置删除成功", nil)
}

type TestConnectionInput struct {
	Email    string `json:"email" binding:"required"`
	IMAPHost string `json:"imap_host" binding:"required"`
	IMAPPort int    `json:"imap_port"`
	Password string `json:"password" binding:"required"`
}

func (h *EmailHandler) TestConnection(c *gin.Context) {
	var input TestConnectionInput
	if err := c.ShouldBindJSON(&input); err != nil {
		utils.Error(c, 400, "请提供完整的连接信息", err)
		return
	}

	port := input.IMAPPort
	if port == 0 {
		port = 993
	}

	success, message := h.emailService.TestConnection(input.Email, input.IMAPHost, port, input.Password)
	c.JSON(200, gin.H{
		"success": success,
		"message": message,
	})
}

func (h *EmailHandler) GetLogs(c *gin.Context) {
	configID := c.Query("configId")
	limit := 50
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil {
			limit = parsed
		}
	}

	ctx, cancel := withReadTimeout(c)
	defer cancel()

	logs, err := h.emailService.GetLogsCtx(ctx, middleware.GetEffectiveUserID(c), configID, limit)
	if err != nil {
		if handleReadTimeoutError(c, err) {
			return
		}
		utils.Error(c, 500, "获取邮件日志失败", err)
		return
	}

	utils.SuccessData(c, logs)
}

func (h *EmailHandler) ClearLogs(c *gin.Context) {
	configID := c.Query("configId")
	if configID == "" {
		utils.Error(c, 400, "missing configId", nil)
		return
	}

	deleted, err := h.emailService.ClearLogs(middleware.GetEffectiveUserID(c), configID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			utils.Error(c, 404, "email config not found", nil)
			return
		}
		utils.Error(c, 500, "failed to clear logs", err)
		return
	}

	utils.Success(c, 200, "logs cleared", gin.H{"deleted": deleted})
}

func (h *EmailHandler) ParseLog(c *gin.Context) {
	id := c.Param("id")
	ownerUserID := middleware.GetEffectiveUserID(c)
	invoice, err := h.emailService.ParseEmailLogCtx(c.Request.Context(), ownerUserID, id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			utils.Error(c, 404, "邮件日志不存在", nil)
			return
		}
		utils.Error(c, 500, "解析邮件发票失败", err)
		return
	}
	utils.Success(c, 200, "解析成功", invoice)
}

func (h *EmailHandler) ExportLogEML(c *gin.Context) {
	id := c.Param("id")
	ownerUserID := middleware.GetEffectiveUserID(c)

	ctx, cancel := withReadTimeout(c)
	defer cancel()

	format := c.Query("format")
	if format == "debug" {
		debug, err := h.emailService.ExportEmailLogDebugCtx(ctx, ownerUserID, id)
		if err != nil {
			if handleReadTimeoutError(c, err) {
				return
			}
			if errors.Is(err, gorm.ErrRecordNotFound) {
				utils.Error(c, 404, "邮件日志不存在", nil)
				return
			}
			utils.Error(c, 500, "导出邮件调试信息失败", err)
			return
		}
		utils.SuccessData(c, debug)
		return
	}

	filename, emlBytes, err := h.emailService.ExportEmailLogEMLCtx(ctx, ownerUserID, id)
	if err != nil {
		if handleReadTimeoutError(c, err) {
			return
		}
		if errors.Is(err, gorm.ErrRecordNotFound) {
			utils.Error(c, 404, "邮件日志不存在", nil)
			return
		}
		utils.Error(c, 500, "导出邮件失败", err)
		return
	}

	if format == "text" {
		c.Header("Content-Type", "text/plain; charset=utf-8")
		c.Header("Content-Disposition", fmt.Sprintf("inline; filename=%q", filename))
		c.Writer.WriteHeader(200)
		_, _ = c.Writer.Write(emlBytes)
		return
	}

	c.Header("Content-Type", "message/rfc822")
	c.Header("Content-Disposition", fmt.Sprintf("attachment; filename=%q", filename))
	c.Writer.WriteHeader(200)
	_, _ = c.Writer.Write(emlBytes)
}

func (h *EmailHandler) StartMonitoring(c *gin.Context) {
	id := c.Param("id")
	started := h.emailService.StartMonitoring(middleware.GetEffectiveUserID(c), id)
	if !started {
		utils.Error(c, 400, "无法启动监控，请检查配置", nil)
		return
	}

	utils.Success(c, 200, "邮箱监控已启动", nil)
}

func (h *EmailHandler) StopMonitoring(c *gin.Context) {
	id := c.Param("id")
	if _, err := h.emailService.GetConfigByID(middleware.GetEffectiveUserID(c), id); err != nil {
		utils.Error(c, 404, "email config not found", nil)
		return
	}
	stopped := h.emailService.StopMonitoring(id)
	if stopped {
		utils.Success(c, 200, "邮箱监控已停止", nil)
	} else {
		utils.Success(c, 200, "监控未在运行", nil)
	}
}

func (h *EmailHandler) GetMonitoringStatus(c *gin.Context) {
	ctx, cancel := withReadTimeout(c)
	defer cancel()

	statuses, err := h.emailService.GetMonitoringStatusCtx(ctx, middleware.GetEffectiveUserID(c))
	if err != nil {
		if handleReadTimeoutError(c, err) {
			return
		}
		utils.Error(c, 500, "获取监控状态失败", err)
		return
	}

	utils.SuccessData(c, statuses)
}

func (h *EmailHandler) ManualCheck(c *gin.Context) {
	id := c.Param("id")
	full := c.Query("full") == "1" || c.Query("full") == "true"
	success, message, newEmails := h.emailService.ManualCheck(middleware.GetEffectiveUserID(c), id, full)

	c.JSON(200, gin.H{
		"success": success,
		"message": message,
		"data": gin.H{
			"newEmails": newEmails,
		},
	})
}
